core/corehttp/hostname.go

```diff
@@ -191,13 +193,36 @@ func HostnameOption() ServeOption {
 							}
 						}
 					}
+				} else { // rootID is not a CID..
+
+					// Check if rootID is a single DNS label with an inlined
+					// DNSLink FQDN a single DNS label. We support this so
+					// loading DNSLink names over TLS "just works" on public
+					// HTTP gateways.
+					//
+					// Rationale for doing this can be found under "Option C"
+					// at: https://github.com/ipfs/in-web-browsers/issues/169
+					//
+					// TLDR is:
+					// https://dweb.link/ipns/my.v-long.example.com
+					// can be loaded from a subdomain gateway with a wildcard
+					// TLS cert if represented as a single DNS label:
+					// https://my-v--long-example-com.ipns.dweb.link
+					if ns == "ipns" && !strings.Contains(rootID, ".") {
+						// my-v--long-example-com → my.v-long.example.com
+						dnslinkFQDN := toDNSLinkFQDN(rootID)
+						if isDNSLinkName(r.Context(), coreAPI, dnslinkFQDN) {
+							// update path prefix to use real FQDN with DNSLink
+							pathPrefix = "/ipns/" + dnslinkFQDN
+						}
+					}
```

Comment:
TODO: resolve original rootID, and if there is no DNSLink, then try dnslinkFQDN one

Reply:
Done: https://github.com/ipfs/go-ipfs/pull/7847/commits/f932510b88a08bb674e4191431b3d33a7b3ec690